---
title: "Covariate analysis of CRANIO counts (original cohort with CQN normalisation and covariates adjusted)"
author: "Thanneer Perumal"
date: "`r date()`"
output: html_document
---
```{r knit2synapse, eval=FALSE}
library(synapseClient)
library(knit2synapse) # get the package from devtools::install_github('Sage-Bionetworks/knit2synapse')

synapseLogin()

knit2synapse::knitToFolderEntity(file = "./covariatesAnalysis_Cohort1.Rmd",
                                 parentId = "syn2820780",
                                 entityName = "Case control and subtype differential expression (original cohort)")
```

```{r libs, echo=FALSE, warning=FALSE, message=FALSE, include=FALSE}
## It is assumed your working directory is where this file

## Load required libraries
library(CovariateAnalysis) # get the package from devtools::install_github('th1vairam/CovariateAnalysis@dev')
library(data.table)
library(tidyr)
library(plyr)
library(dplyr)
library(stringr)

library(ggplot2)
library(reshape2)
library(limma)
library(gplots)
library(WGCNA)
library(psych)
library(edgeR)
library(biomaRt)
library(RColorBrewer)
library(cqn)

library(synapseClient)
library(knitr)
library(githubr) # get the package from devtools::install_github('brian-bot/githubr')

synapseLogin()

library(doParallel)
library(foreach)

cl = makeCluster(detectCores()-2)
registerDoParallel(cl)

options(xtable.type="html")

knitr::opts_chunk$set(
  echo=FALSE,
  warning=FALSE,
  message=FALSE,
  error = FALSE,
  tidy = FALSE,
  cache = TRUE)
```

```{r synapse.parameters, include=FALSE, cache=TRUE}
# Synapse parameters
parentId = 'syn2820780';
activityName = 'Covariate analysis of validation cohort';
activityDescription = 'Covariate analysis of RNAseq data from craniosynostosis validation cohort';

thisFileName <- 'covariatesAnalysis_Cohort1.Rmd'

# Github link
thisRepo <- getRepo(repository = "th1vairam/CRANIO", ref="branch", refName='diffexp')
thisFile <- getPermlink(repository = thisRepo, repositoryPath = thisFileName)
```
### Data download
#### Obtain count matrix and metadata from synapse.
```{r download.data, cache=TRUE}
# Download expression data
COUNT_ID <- 'syn2820309';
ALL_USED_IDs <- COUNT_ID
COUNT_OBJ <- synGet(COUNT_ID)
COUNT <- read.table(COUNT_OBJ@filePath,header=T,sep=',',check.names = F, row.names = 1)

# Get picard metrics from synapse
METADATA.PICARD_ID <- 'syn2820392'
ALL_USED_IDs[length(ALL_USED_IDs)+1] = METADATA.PICARD_ID
METADATA.PICARD <- downloadFile(METADATA.PICARD_ID) %>%
  plyr::rename(c('SAMPLE' = 'SampleID'))

# Get metadata
METADATA_ID <- 'syn3241418'
ALL_USED_IDs[length(ALL_USED_IDs)+1] = METADATA_ID
METADATA <- downloadFile(METADATA_ID) %>%
  plyr::rename(c('SeqSampleName' = 'SampleID')) %>%
  dplyr::select(SampleID, Age_mos., Sex, Sample_Type, Initial_growth_duration_days, Days_in_LN2,
                RNA_days_to_harvest, Sample_Origin, libraryBatch) %>%
  dplyr::left_join(METADATA.PICARD)

# Get gene specific parameters from synapse
GENE.PARAM = downloadFile('syn8449369')
ALL_USED_IDs[length(ALL_USED_IDs)+1] = 'syn8449369'

GENE.LEN = dplyr::select(GENE.PARAM, ensembl_gene_id, gene.length) %>% unique()
rownames(GENE.LEN) = GENE.LEN$ensembl_gene_id

GENE.GC = dplyr::select(GENE.PARAM, ensembl_gene_id, percentage_gc_content) %>% unique()
rownames(GENE.GC) = GENE.GC$ensembl_gene_id
```

### Data preprocessing
```{r preprocess.data,cache=TRUE, echo=TRUE}
# Remove samples with no Sex and Diagnosis information
METADATA <- METADATA %>%
  filter(!is.na(Sex), !is.na(Sample_Type), !is.na(SampleID),
         !is.na(Age_mos.)) %>%
  dplyr::mutate(Diagnosis = factor(Sample_Type,
                                   levels = c("Control", "Coronal Synostosis L", "Coronal Synostosis R", 
                                              "Lambdoid", "Metopic Synostosis", "Sagittal Synostosis"),
                                   labels = c("Control" = "Control", "Coronal Synostosis L" = "Coronal", 
                                              "Coronal Synostosis R" = "Coronal", "Lambdoid" = "Lambdoid", 
                                              "Metopic Synostosis" = "Metopic", "Sagittal Synostosis" = "Sagittal")),
                Sample_Origin = gsub('[^[:alnum:]]','_', Sample_Origin)) %>%
  droplevels() %>%
  as.data.frame()
```

```{r preprocess.data1,cache=TRUE}
# Match covariates to expression data
indToRetain = intersect(METADATA$SampleID, colnames(COUNT))
removedIDs = setdiff(colnames(COUNT), METADATA$SampleID)

COUNT = COUNT[,indToRetain]

rownames(METADATA) = METADATA$SampleID
METADATA = METADATA[indToRetain,]
```
`r dim(COUNT)[2]` samples from the validation cohorts are used for the analysis. Following sample are removed `r paste(removedIDs, collapse = ',')` from analysis for missing key metadata

### Covariate clustering
Determine relationship between covariates
```{r covariates.clustering, cache=TRUE}
primaryVariable <- "Diagnosis"
FactorCovariates <- c("Diagnosis", "Sex", "Sample_Origin", "libraryBatch")
ContCovariates <- c("Age_mos.", "Initial_growth_duration_days", "Days_in_LN2", "RNA_days_to_harvest",
                    "MEDIAN_5PRIME_TO_3PRIME_BIAS", "MEDIAN_CV_COVERAGE", "PCT_CODING_BASES", "PCT_CORRECT_STRAND_READS",
                    "PCT_INTERGENIC_BASES", "PCT_INTRONIC_BASES", "PCT_MRNA_BASES", "PCT_UTR_BASES",
                    "PF_ALIGNED_BASES", "PF_BASES")

# Find inter relation between factor covariates
COVARIATES = METADATA[,c(FactorCovariates,ContCovariates),drop=F]
rownames(COVARIATES) <- METADATA$SampleID

# Convert factor covariates to factors
COVARIATES[,FactorCovariates] = lapply(COVARIATES[,FactorCovariates], factor)
COVARIATES[,ContCovariates] = lapply(COVARIATES[,ContCovariates], as.character)
COVARIATES[,ContCovariates] = lapply(COVARIATES[,ContCovariates], as.numeric)
```
Correlation/association between covariates at an FDR <= 0.1
```{r covariates.correlation, fig.width=6, fig.height=6}
COVARIATES.CORRELATION = getAssociationStatistics(COVARIATES, PVAL = 0.05)
draw(COVARIATES.CORRELATION$plot, heatmap_legend_side = 'left')
```

### CPM Normalisation
Preprocess counts matrix using CQN (Conditional Quantile Normlaisation accounting for both GC content and gene length)

Remove genes that have less than 1 cpm counts in at least 50% of samples per Diagnosis. Also remove genes with missing gene length and percentage GC content
```{r cpmnormalisation}
genesToAnalyze = COVARIATES %>%
  rownameToFirstColumn('Sample_Name') %>%
  dlply(.(Diagnosis), .fun = function(mtd, count){
    processed.counts = getGeneFilteredGeneExprMatrix(count[,mtd$Sample_Name],
                                                     MIN_GENE_CPM=1, 
                                                     MIN_SAMPLE_PERCENT_WITH_MIN_GENE_CPM=0.5)
    processed.counts$filteredExprMatrix$genes
  }, COUNT)

genesToAnalyze = unlist(genesToAnalyze) %>% 
  unique() %>%
  intersect(GENE.GC$ensembl_gene_id[!is.na(GENE.GC$percentage_gc_content)]) %>%
  intersect(GENE.LEN$ensembl_gene_id[!is.na(GENE.LEN$gene.length)]) %>%
  setdiff(c("N_unmapped", "N_multimapping", "N_noFeature", "N_ambiguous"))

PROCESSED_COUNTS = getGeneFilteredGeneExprMatrix(COUNT[genesToAnalyze, ], 
                                                 MIN_GENE_CPM=0, 
                                                 MIN_SAMPLE_PERCENT_WITH_MIN_GENE_CPM=0)
```
Processing `r dim(PROCESSED_COUNTS$filteredExprMatrix)[1]` genes in `r dim(PROCESSED_COUNTS$filteredExprMatrix)[2]` samples

### Outlier analysis
Clustering of logCPM (unadjusted)
```{r decompse.normalise.data1, fig.height=6, fig.width=8, results='asis'}
# Find principal components of expression to plot
EXPR = voom(PROCESSED_COUNTS$filteredExprMatrix$counts)
PC <- prcomp(EXPR$E, scale.=T, center = T)

# Plot first 2 PCs
plotdata <- data.frame(SampleID=rownames(PC$rotation), 
                       PC1=PC$rotation[,1], 
                       PC2=PC$rotation[,2])

plotdata <- left_join(plotdata, rownameToFirstColumn(COVARIATES, 'SampleID')) %>%
  dplyr::mutate(label = SampleID)

p <- ggplot(plotdata, aes(x=PC1, y=PC2))
p <- p + geom_point(aes(color=libraryBatch, shape=Diagnosis))
p <- p + theme_bw() + theme(legend.position="right")
# p <- p + geom_text(aes(label= label), size=4, hjust=1)
p
```

```{r remove.samples, fig.height=6, fig.width=8, results='asis'}
indToRemove = c('')
indToRetain = setdiff(rownames(COVARIATES), indToRemove)
NEW.COUNTS = PROCESSED_COUNTS$filteredExprMatrix$counts[, indToRetain]
COVARIATES = COVARIATES[indToRetain,]
```
Removing following samples as outliers `r paste(indToRemove, collapse = ', ')`

### Normalisation (with NULL)
Initial normalisation usign cqn and dispersion control with voom (with NULL design)
```{r initial.voom.normalisation, fig.height=5, fig.width=5}
# Compute offset for gene length and gc content
CQN.GENE_EXPRESSION = cqn(NEW.COUNTS, 
                          x = GENE.GC[rownames(NEW.COUNTS), 'percentage_gc_content'],
                          lengths = GENE.LEN[rownames(NEW.COUNTS), 'gene.length'],
                          lengthMethod = "smooth", 
                          verbose = FALSE)

# Set gene counts in specific samples that are deviating 3 sd from other samples to 0
log.mat = apply(CQN.GENE_EXPRESSION$y+CQN.GENE_EXPRESSION$offset, 1, function(x){
  mn = mean(x, na.rm = T)
  std.dev = sd(x, na.rm = T)
  return((x < (mn-3*std.dev)) | (x > (mn+3*std.dev)))
}) %>% t
NEW.COUNTS[log.mat] = 0

# Re-compute offset for gene length and gc content
CQN.GENE_EXPRESSION = cqn(NEW.COUNTS, 
                          x = GENE.GC[rownames(NEW.COUNTS), 'percentage_gc_content'],
                          lengths = GENE.LEN[rownames(NEW.COUNTS), 'gene.length'],
                          lengthMethod = "smooth", 
                          verbose = FALSE)

# Plot distribution of expression
tmp = (CQN.GENE_EXPRESSION$y+CQN.GENE_EXPRESSION$offset) %>%
  rownameToFirstColumn('ensembl_gene_id') %>%
  gather(SampleID, Value, -ensembl_gene_id) 
p = ggplot(tmp, aes(color = SampleID, x = Value)) + geom_density() + xlab('logCPM') + ylab('Density')
p = p + theme(legend.position = 'NONE')
p
```
Coexpression of genes 
```{r coexp1, cache=FALSE, fig.height=5, fig.width=5}
cr = cor(t(CQN.GENE_EXPRESSION$y+CQN.GENE_EXPRESSION$offset))
hist(cr, main = 'Distribution of correlation between genes', xlab = 'Correlation')
```
### Expression decomposition
Clustering of logcpm+offset
```{r decompse.normalise.data1.1, fig.height=6, fig.width=8, results='asis'}
# Find principal components of expression to plot
PC <- prcomp(CQN.GENE_EXPRESSION$y+CQN.GENE_EXPRESSION$offset, scale.=T, center = T)

# Plot first 2 PCs
plotdata <- data.frame(SampleID=rownames(PC$rotation), 
                       PC1=PC$rotation[,1], 
                       PC2=PC$rotation[,2])

plotdata <- left_join(plotdata, rownameToFirstColumn(COVARIATES, 'SampleID'))

p <- ggplot(plotdata, aes(x=PC1, y=PC2))
p <- p + geom_point(aes(color=libraryBatch, shape=Diagnosis))
p <- p + theme_bw() + theme(legend.position="right")
# p <- p + geom_text(aes(label= SampleID), size=4, hjust=1)
p
```
Tree based classification of samples
```{r decompse.normalise.data1.2, fig.height=6, fig.width=10, results='asis'}
# Eucledian tree based analysis
COVARIATES.tmp = (COVARIATES[,c(FactorCovariates), drop = F])
COVARIATES.tmp[is.na(COVARIATES.tmp)] = 0

tree = hclust(as.dist(t(CQN.GENE_EXPRESSION$y + CQN.GENE_EXPRESSION$offset)))
cols = WGCNA::labels2colors(COVARIATES.tmp);

WGCNA::plotDendroAndColors(tree, 
                           colors = cols, 
                           dendroLabels = FALSE, 
                           abHeight = 0.80, 
                           main = "Sample dendrogram",
                           groupLabels = colnames(COVARIATES.tmp))
```

```{r temp, include = F}
dev.off()
gc()
```

### Significant Covariates
Correlation between pca of unadjusted mRNA expression and covariates are used to find significant covariates
```{r preadj.covariates, cache=TRUE}
# Find correlation between PC's of gene expression with covariates
preAdjustedSigCovars = runPCAandPlotCorrelations(CQN.GENE_EXPRESSION$y+CQN.GENE_EXPRESSION$offset, 
                                                 COVARIATES,
                                                 'NULL design(voom-normalized)', 
                                                 isKeyPlot=TRUE, 
                                                 MIN_PVE_PCT_PC = 1)
```

Significant covariates to adjust at FDR 0.1 are `r preAdjustedSigCovars$significantCovars`
```{r preadj.covariates.plot, fig.width=12, fig.height=8}
preAdjustedSigCovars[["PC_res"]][[2]]$plotData
```
### Normalisation (iterative design)
Since many covariates are correlated, re-normalising and re-adjusting COUNTS with an iterative design matrix.
1. Adding libraryBatch and Sex a priori to variable selection
2. Primary variable of interest "Diagnosis"" is excluded from the pool of available covariates for selection
```{r iterative.norm, results='asis'}
# Primary variable of interest
postAdjustCovars = c('libraryBatch', 'Sex');

# Assign residual covariates
residualCovars = setdiff(preAdjustedSigCovars$significantCovars, c(postAdjustCovars, primaryVariable))
residualSigCovars = preAdjustedSigCovars
covariatesEffects = preAdjustedSigCovars$Effects.significantCovars[residualCovars]
postAdjustCovars = c(postAdjustCovars, names(which.max(covariatesEffects))) %>% unique()

loopCount = 0 
while(length(residualSigCovars$significantCovars)!=0 && loopCount <= 20){
  writeLines(paste('Using following covariates in the model:',
                   paste(postAdjustCovars, collapse=', '),
                   'as fixed effects'))
  
  # Post adjusted design matrix
  DM1 = getDesignMatrix(COVARIATES[,postAdjustCovars,drop=F],Intercept = F)
  DM1$design = DM1$design[,linColumnFinder(DM1$design)$indepCols]
  
  # Estimate voom weights for dispersion control
  VOOM.GENE_EXPRESSION = voom(NEW.COUNTS, design=DM1$design, plot=F)
  
  # Fit linear model using new weights and new design
  VOOM.ADJUSTED.FIT = lmFit(CQN.GENE_EXPRESSION$y+CQN.GENE_EXPRESSION$offset,
                            design = DM1$design,
                            weights = VOOM.GENE_EXPRESSION$weights)
  
  # Residuals after normalisation
  RESIDUAL.GENE_EXPRESSION = residuals.MArrayLM(VOOM.ADJUSTED.FIT,
                                                CQN.GENE_EXPRESSION$y+CQN.GENE_EXPRESSION$offset)
  
  # Residual covariates to choose from
  residCovars <- setdiff(c(FactorCovariates,ContCovariates), c(postAdjustCovars, primaryVariable))
  
  # Find PC of residual gene expression and significant covariates that are highly correlated with PCs
  residualSigCovars = runPCAandPlotCorrelations(RESIDUAL.GENE_EXPRESSION, 
                                                COVARIATES[, residCovars, drop=F], 
                                                'adjusted design(voom-normalized)',
                                                isKeyPlot=TRUE)
  
  # Add postadjusted covariates (if any)
  residCovars = setdiff(residualSigCovars$significantCovars, c(postAdjustCovars, primaryVariable))
  covariatesEffects = residualSigCovars$Effects.significantCovars[residCovars]
  
  postAdjustCovars = c(postAdjustCovars, names(which.max(covariatesEffects)))
  loopCount = loopCount + 1
}

modelStr <- paste(paste(gsub('_','\\\\_',postAdjustCovars), collapse=', '),
                  'as fixed effects')

tmp <- paste('Using following covariates in the final model:', modelStr)
```
`r tmp`

### Sanity check
```{r residual.adj, fig.width=12, fig.height=8}
# Find PC of residual gene expression and significant covariates that are highly correlated with PCs
residualSigCovars = runPCAandPlotCorrelations(RESIDUAL.GENE_EXPRESSION, 
                                              COVARIATES,
                                              'adjusted design(voom-normalized)',
                                              isKeyPlot=TRUE)

residualSigCovars[["PC_res"]][[2]]$plotData
```
Coexpression of genes 
```{r coexp2, cache=FALSE, fig.height=5, fig.width=5}
cr = cor(t(RESIDUAL.GENE_EXPRESSION))
hist(cr, main = 'Distribution of correlation between genes', xlab = 'Correlation')
```

### Clustering residual data
```{r decompse.normalise.data2.1, fig.height=6, fig.width=8, results='asis'}
# Find principal components of expression to plot
PC <- prcomp(RESIDUAL.GENE_EXPRESSION, scale.=T, center = T)

# Plot first 4 PCs
plotdata <- data.frame(SampleID=rownames(PC$rotation), 
                       PC1=PC$rotation[,1], 
                       PC2=PC$rotation[,2])

plotdata <- left_join(plotdata, rownameToFirstColumn(COVARIATES, 'SampleID'))

p <- ggplot(plotdata, aes(x=PC1, y=PC2)) 
p <- p + geom_point(aes(color = Diagnosis))
p <- p + theme_bw() + theme(legend.position="right")
# p <- p + geom_text(aes(label= SampleID), size=4, hjust=1)
p
```

```{r decompse.normalise.data2.2, fig.height=8, fig.width=12, results='asis'}
# Eucledian tree based analysis
COVARIATES.tmp = data.matrix(COVARIATES[,FactorCovariates])
COVARIATES.tmp[is.na(COVARIATES.tmp)] = 0

tree = hclust(as.dist(t(RESIDUAL.GENE_EXPRESSION)))
cols = WGCNA::labels2colors(COVARIATES.tmp);

WGCNA::plotDendroAndColors(tree, 
                           colors = cols, 
                           dendroLabels = FALSE, 
                           abHeight = 0.80, 
                           main = "Sample dendrogram",
                           groupLabels = colnames(COVARIATES.tmp))
```

```{r temp1, include=F}
dev.off()
gc()
```

### Differential expression analysis (between sutre types)
Differential expression is performed on the primary variable by controlling for covariates identified above

Following model was used

Gx = Diagnosis + libraryBatch + Sex + RNA\\_days\\_to\\_harvest + PCT\\_CORRECT\\_STRAND\\_READS + 
      PCT\\_INTERGENIC\\_BASES + Age\\_mos. + Initial\\_growth\\_duration\\_days + MEDIAN\\_CV\\_COVERAGE + 
      PCT\\_INTRONIC\\_BASES + Sample\\_Origin + MEDIAN\\_5PRIME\\_TO\\_3PRIME\\_BIAS

Genes that are differentially expressed at an FDR <= 0.05 are
```{r diff.exp, fig.height=16, fig.width=18, cache=FALSE}
# Get design matrix
DESIGN = getDesignMatrix(COVARIATES[, c(primaryVariable, postAdjustCovars), drop = F], Intercept = F)
DESIGN$design = DESIGN$design[,linColumnFinder(DESIGN$design)$indepCols]

# Estimate voom weights for dispersion control
VOOM.WEIGHTS = voom(NEW.COUNTS, design=DESIGN$design, plot=F)

# Fit linear model using new weights and new design
FIT = lmFit(CQN.GENE_EXPRESSION$y+CQN.GENE_EXPRESSION$offset,
            design = DESIGN$design,
            weights = VOOM.WEIGHTS$weights)

# Fit contrast
contrast = paste0('Diagnosis',levels(COVARIATES$Diagnosis)) %>%
  combn(2) %>%
  apply(2, function(x){
    paste(x, collapse = '-')
  }) %>%
  c('DiagnosisControl-(DiagnosisCoronal+DiagnosisLambdoid+DiagnosisMetopic+DiagnosisSagittal)/4',
    'RNA_days_to_harvest') %>%
  makeContrasts(contrasts=.,
                levels = colnames(FIT$coefficients))
FIT.CONTR = contrasts.fit(FIT, contrasts=contrast)
FIT.CONTR = eBayes(FIT.CONTR)

# Get differnetial expression
DE = lapply(1:11, function(i, FIT){
  topTable(FIT, coef=i, number = 50000, confint = T) %>%
    rownameToFirstColumn('ensembl_gene_id')
}, FIT.CONTR)
names(DE) = colnames(contrast)[1:11]

DE1 = DE %>% 
  rbindlist(idcol = 'Comparison') %>%
  dplyr::mutate(Comparison = gsub('Diagnosis','',Comparison),
                Comparison = gsub('[^[:alnum:]]','_', Comparison),
                Comparison = gsub('_Coronal_Lambdoid_Metopic_Sagittal__4', 'Cranio', Comparison)) %>%
  tidyr::separate(Comparison, c('reference','against'), sep = '_') %>%
  left_join(GENE.PARAM %>%
              dplyr::select(ensembl_gene_id, hgnc_symbol, percentage_gc_content, gene.length) %>%
              unique()) %>%
  dplyr::mutate(Direction = logFC/abs(logFC),
                Direction = factor(Direction, c(-1,1), c('-1' = 'DOWN', '1' = 'UP')),
                Direction = as.character(Direction))
DE1$Direction[DE1$adj.P.Val > 0.05] = 'NONE'

writeLines('Differentially expressed genes at an FDR 0.05')
tmp = DE1 %>%
  dplyr::select(ensembl_gene_id, reference, against, Direction) %>%
  group_by(reference, against, Direction) %>%
  dplyr::summarise(FDR_0_05 = length(unique(ensembl_gene_id))) %>%
  dplyr::filter(Direction != 'NONE') %>%
  spread(Direction, FDR_0_05) 
kable(tmp)

p = ggplot(DE1, aes(y = -log10(adj.P.Val), x = logFC, color = Direction)) + geom_point() + xlim(c(-1,1))
p = p + scale_color_manual(values = c('green','grey','red'))
p = p + facet_grid(reference+.~against, scales = 'fixed')
p
```
### Differential expression wrt RNA days to harvest
```{r rna.harvest}
# Get differnetial expression
DE2 = topTable(FIT.CONTR, coef=12, number = 50000, confint = T) %>%
  rownameToFirstColumn('ensembl_gene_id') %>%
  dplyr::mutate(reference = 'RNA_days_to_Harvest') %>%
  left_join(GENE.PARAM %>%
              dplyr::select(ensembl_gene_id, hgnc_symbol, percentage_gc_content, gene.length) %>%
              unique()) %>%
  dplyr::mutate(Direction = logFC/abs(logFC),
                Direction = factor(Direction, c(-1,1), c('-1' = 'DOWN', '1' = 'UP')),
                Direction = as.character(Direction))
DE2$Direction[DE2$adj.P.Val > 0.05] = 'NONE'

writeLines('Differentially expressed genes at an FDR 0.05')
tmp = DE2 %>%
  dplyr::select(ensembl_gene_id, reference, Direction) %>%
  group_by(reference, Direction) %>%
  dplyr::summarise(FDR_0_05 = length(unique(ensembl_gene_id))) %>%
  dplyr::filter(Direction != 'NONE') %>%
  spread(Direction, FDR_0_05) 
kable(tmp)

p = ggplot(DE2, aes(y = -log10(adj.P.Val), x = logFC, color = Direction)) + geom_point() + xlim(c(-1,1))
p = p + scale_color_manual(values = c('green','grey','red')) + theme(legend.position = 'top')
p = p + facet_grid(.~reference, scales = 'fixed')
p

DE = rbindlist(list(DE1, DE2), use.names = T, fill = T)
```

### Associate differential expression results with gc content, gene length and average expression
```{r associate.de, fig.height=6, fig.width=15}
pl = list()
pl[[1]] = ggplot(DE1 %>% dplyr::filter(reference == 'Control', against == 'Cranio'), 
                 aes(x = log10(gene.length), y = logFC, color = Direction)) + geom_point() 
pl[[1]] = pl[[1]] + geom_smooth(method = 'loess', inherit.aes = FALSE, aes(x = log10(gene.length), y = logFC))
pl[[1]] = pl[[1]] + facet_grid(reference~.+against) + scale_color_manual(values = c('green', 'grey', 'red'))
pl[[1]] = pl[[1]] + theme(legend.position = 'top')

pl[[2]] = ggplot(DE1 %>% dplyr::filter(reference == 'Control', against == 'Cranio'), 
                 aes(x = percentage_gc_content, y = logFC, color = Direction)) + geom_point()
pl[[2]] = pl[[2]] + geom_smooth(method = 'loess', inherit.aes = FALSE, aes(x = percentage_gc_content, y = logFC))
pl[[2]] = pl[[2]] + facet_grid(reference~.+against) + scale_color_manual(values = c('green', 'grey', 'red'))
pl[[2]] = pl[[2]] + theme(legend.position = 'top')

pl[[3]] = ggplot(DE1 %>% dplyr::filter(reference == 'Control', against == 'Cranio'), 
                 aes(x = AveExpr, y = logFC, color = Direction)) + geom_point()
pl[[3]] = pl[[3]] + geom_smooth(method = 'loess', inherit.aes = FALSE, aes(x = AveExpr, y = logFC))
pl[[3]] = pl[[3]] + facet_grid(reference~.+against) + scale_color_manual(values = c('green', 'grey', 'red'))
pl[[3]] = pl[[3]] + theme(legend.position = 'top')

multiplot(plotlist = pl, cols = 3)
```

### Store files in synapse
```{r synapse.store, include=FALSE, eval=TRUE, cache=FALSE}
# Set annotations
all.annotations = list(
  dataType = 'mRNA',
  dataSubType = 'geneExp',
  summaryLevel = 'gene',
  assay	 = 'RNAseq',
  
  study = 'CRANIO1', 

  organism = 'HomoSapiens',

  normalizationStatus	= TRUE,
  modelSystem	= FALSE,
  
  normalizationType	= 'CQN'
)

# Code
CODE <- Folder(name = "Case control and subtype differential expression (original cohort)", parentId = parentId)
annotations(CODE) = all.annotations
CODE <- synStore(CODE)

# Store covariates
COVARIATES = rownameToFirstColumn(COVARIATES, 'SampleID')
write.table(COVARIATES, file = 'CRANIO1_Covariates.tsv', sep = '\t', row.names=F, quote=F)
COV_OBJ = File('CRANIO1_Covariates.tsv', name = 'Covariates', parentId = CODE$properties$id)
annotations(COV_OBJ) = annotations(CODE)
annotations(COV_OBJ)$dataSubType = 'covariates'
COV_OBJ = synStore(COV_OBJ, used = ALL_USED_IDs, activityName = activityName, 
                   executed = thisFile, activityDescription = activityDescription)

# Store filtered counts
PROCESSED_COUNTS$filteredExprMatrix$counts %>%
  rownameToFirstColumn('ensembl_gene_id') %>%
  write.table(file = 'CRANIO1_Counts.tsv', sep = '\t', row.names=F, quote=F)
COUNT_OBJ = File('CRANIO1_Counts.tsv', name = 'Counts (filtered raw)', parentId = CODE$properties$id)
annotations(COUNT_OBJ) = annotations(CODE)
annotations(COUNT_OBJ)$dataSubType = 'filteredCounts'
COUNT_OBJ = synStore(COUNT_OBJ, activity = synGetActivity(COV_OBJ$properties$id))

# Store logCPM
CQN.GENE_EXPRESSION$y %>%
  rownameToFirstColumn('ensembl_gene_id') %>%
  write.table(file = 'CRANIO1_logCPM.tsv', sep = '\t', row.names=F, quote=F)
LCOUNT_OBJ = File('CRANIO1_logCPM.tsv', name = 'Counts (filtered logCPM)', parentId = CODE$properties$id)
annotations(LCOUNT_OBJ) = annotations(CODE)
annotations(LCOUNT_OBJ)$dataSubType = 'filteredLCPM'
LCOUNT_OBJ = synStore(LCOUNT_OBJ, activity = synGetActivity(COV_OBJ$properties$id))

# Store cqn offsets
CQN.GENE_EXPRESSION$offset %>%
  rownameToFirstColumn('ensembl_gene_id') %>%
  write.table(file = 'CRANIO1_offset.tsv', sep = '\t', row.names=F, quote=F)
OFFSET_OBJ = File('CRANIO1_offset.tsv', name = 'Gene length and GC content offset', parentId = CODE$properties$id)
annotations(OFFSET_OBJ) = annotations(CODE)
annotations(OFFSET_OBJ)$dataSubType = 'offset'
OFFSET_OBJ = synStore(OFFSET_OBJ, activity = synGetActivity(COV_OBJ$properties$id))

# Store design matrix
DESIGN$design %>%
  rownameToFirstColumn('SampleID') %>%
  write.table(file = 'CRANIO1_Design.tsv', sep = '\t', row.names=F, quote=F)
DM_OBJ = File('CRANIO1_Design.tsv', name = 'Design Matrix', parentId = CODE$properties$id)
annotations(DM_OBJ) = annotations(CODE)
annotations(DM_OBJ)$dataSubType = 'designMatrix'
DM_OBJ = synStore(DM_OBJ, activity = synGetActivity(COV_OBJ$properties$id))

# Store residual gene expression for network analysis
RESIDUAL.GENE_EXPRESSION %>%
  rownameToFirstColumn('ensembl_gene_id') %>%
  write.table(file = 'CRANIO1_netResidualExpression.tsv', sep = '\t', row.names=F, quote=F)
nEXP_OBJ = File('CRANIO1_netResidualExpression.tsv', 
                name = 'Normalised, covariates removed residual expression (for network analysis)', 
                parentId = CODE$properties$id)
annotations(nEXP_OBJ) = annotations(CODE)
annotations(nEXP_OBJ)$dataSubType = 'residualGeneExpForNetAnlz'
nEXP_OBJ = synStore(nEXP_OBJ, activity = synGetActivity(COV_OBJ$properties$id))

# Store differential expression results
rbindlist(list(DE, DE1), use.names = T, fill = T) %>%
  write.table(file = 'CRANIO1_DiffExpression.tsv', sep = '\t', row.names=F, quote=F)
DEXP_OBJ = File('CRANIO1_DiffExpression.tsv', name = 'Differential Expression Results', 
                parentId = CODE$properties$id)
annotations(DEXP_OBJ) = annotations(CODE)
annotations(DEXP_OBJ)$dataSubType = 'diffExp'
DEXP_OBJ = synStore(DEXP_OBJ, activity = synGetActivity(COV_OBJ$properties$id))
```
|  *Results*                                  |  *SynapseID*                     |
|  -----------------------------------------  |   ---------                      |
|  Covariates                                 |  `r COV_OBJ$properties$id`       |
|  Counts (raw)                               |  `r COUNT_OBJ$properties$id`     |
|  Counts (lcpm)                              |  `r LCOUNT_OBJ$properties$id`    |
|  Offset (for gene length and gc content)    |  `r OFFSET_OBJ$properties$id`    |
|  Design Matrix                              |  `r DM_OBJ$properties$id`        |
|  Residual Expression (for network analysis) |  `r nEXP_OBJ$properties$id`      |
|  Differential Expression                    |  `r DEXP_OBJ$properties$id`      |